stats/statsd/common/collector.go

```diff
@@ -0,0 +1,199 @@
+	// of those tags that should be sent. No tags are send in case of ProcessTags being null
+	ProcessTags func(map[string]string) []string
+
+	logger     *log.Entry
+	client     *statsd.Client
+	startTime  time.Time
+	buffer     []*Sample
+	bufferLock sync.Mutex
+}
+
+// Init sets up the collector
+func (c *Collector) Init() (err error) {
+	c.logger = log.WithField("type", c.Type)
+	if address := c.Config.Addr.String; address == "" {
+		err = fmt.Errorf(
+			"connection string is invalid. Received: \"%+s\"",
+			address,
+		)
+		c.logger.Error(err)
+
+		return err
+	}
+
+	c.client, err = statsd.NewBuffered(c.Config.Addr.String, int(c.Config.BufferSize.Int64))
+
+	if err != nil {
+		c.logger.Errorf("Couldn't make buffered client, %s", err)
+		return err
+	}
+
+	if namespace := c.Config.Namespace.String; namespace != "" {
+		c.client.Namespace = namespace
+	}
+
+	return nil
+}
+
+// Link returns the address of the client
+func (c *Collector) Link() string {
+	return c.Config.Addr.String
+}
+
+// Run the collector
+func (c *Collector) Run(ctx context.Context) {
+	c.logger.Debugf("%s: Running!", c.Type)
+	ticker := time.NewTicker(time.Duration(c.Config.PushInterval.Duration))
+	c.startTime = time.Now()
+
+	for {
+		select {
+		case <-ticker.C:
+			c.pushMetrics()
+		case <-ctx.Done():
+			c.pushMetrics()
+			c.finish()
+			return
+		}
+	}
+}
+
+// GetRequiredSystemTags Return the required system sample tags for the specific collector
+func (c *Collector) GetRequiredSystemTags() lib.TagSet {
+	return lib.TagSet{} // no tags are required
+}
+
+// SetRunStatus does nothing in statsd collector
+func (c *Collector) SetRunStatus(status lib.RunStatus) {}
+
+// Collect metrics
+func (c *Collector) Collect(containers []stats.SampleContainer) {
+	var pointSamples []*Sample
+
+	for _, container := range containers {
+		for _, sample := range container.GetSamples() {
+			pointSamples = append(pointSamples, generateDataPoint(sample))
+		}
+	}
+
+	if len(pointSamples) > 0 {
+		c.bufferLock.Lock()
+		c.buffer = append(c.buffer, pointSamples...)
+		c.bufferLock.Unlock()
+	}
+}
+
+func (c *Collector) pushMetrics() {
+	c.bufferLock.Lock()
+	if len(c.buffer) == 0 {
+		c.bufferLock.Unlock()
+		return
+	}
+	buffer := c.buffer
+	c.buffer = nil
+	c.bufferLock.Unlock()
+
+	c.logger.
+		WithField("samples", len(buffer)).
+		Debug("Pushing metrics to server")
+
+	if err := c.commit(buffer); err != nil {
+		c.logger.
+			WithError(err).
+			Error("Couldn't commit a batch")
+	}
+}
+
+func (c *Collector) finish() {
+	// Close when context is done
+	if err := c.client.Close(); err != nil {
+		c.logger.Warnf("Error closing the client, %+v", err)
+	}
+}
+
+func (c *Collector) commit(data []*Sample) error {
+	for _, entry := range data {
+		if err := c.dispatch(entry); err != nil {
+			// No need to return error if just one metric didn't go through
+			c.logger.WithError(err).Warnf("Error while sending metric %s", entry.Metric)
```

Comment:
I think that this will lead to people wondering where their metrics are and it will make it harder to debug. If people complain that it is way too spammy in situations which are okay to not be spammy we can always make it less spammy

Reply:
Come on, this line can generate literally thousands of warning messages - not very useful. I'm not saying to suppress those errors, just to aggregate them and say "1234 metrics weren't sent" every 1 second, instead of 1234 messages every second...

Reply:
my argument is that knowing what didn't get sent may help you debug it ... hiding the error behind why we couldn't send something and just telling you we couldn't will lead to questions of "why do I have no metrics?" and "what does 'we couldn't sent your metrics' mean?".

If someone is continuously getting 1234 messages, that we could not sent the data, they told us to sent, the error message *might* help them debug it. In case someone doesn't care that they can't sent the majority of they are metrics ... maybe they should not ask us to do it ?

Reply:
Read my original comment again - I don't want to totally suppress the messages, just aggregate them by default. If they run k6 with verbose mode enabled, users will see the original error messages, however many there are. We can even suggest turning on the verbose mode in the aggregated error message. Or we can show the first 5 or 10 error messages and just mention that there are 1229 more :smile: 

I'm fine with a lot of different variants, as long as we don't bury the user in thousands of (likely similar) error messages on the console every second by default, which is terrible UX.